xwiki-commons-core/xwiki-commons-configuration/xwiki-commons-configuration-default/src/main/java/org/xwiki/configuration/internal/AbstractSystemOverwriteConfigurationSource.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.configuration.internal;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import jakarta.inject.Inject;
+
+import org.xwiki.component.descriptor.ComponentDescriptor;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.configuration.ConfigurationSource;
+
+/**
+ * Helper to implement {@link ConfigurationSource}.
+ * 
+ * @version $Id$
+ * @since 17.5.0RC1
+ */
+public abstract class AbstractSystemOverwriteConfigurationSource extends AbstractConfigurationSource
+    implements Initializable
+{
+    @Inject
+    protected ComponentDescriptor<AbstractSystemOverwriteConfigurationSource> componentDescriptor;
+
+    @Inject
+    protected ComponentManager componentManager;
+
+    /**
+     * True if it should be possible to overwrite this configuration source using system environment variables or
+     * properties.
+     */
+    protected boolean systemOverwriteEnabled;
+
+    protected ConfigurationSource systemConfigurationSource;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        if (this.systemOverwriteEnabled) {
+            try {
+                this.systemConfigurationSource = this.componentManager.getInstance(ConfigurationSource.class, "system");
+            } catch (ComponentLookupException e) {
+                throw new InitializationException("Failed to lookup the system configuration source", e);
+            }
+        }
+    }
+
+    /**
+     * @param key the key
+     * @return the key to use with the system ConfigurationSource
```

Comment:
not so clear for the javadoc why `key` is what.

Reply:
Yes, the javadoc is light, but it's honestly almost a private thing actually (I just let it be `protected` just in case). I add some more stuff.